### PR TITLE
Fixing docker images and adding test for sagemaker_sdk.

### PR DESCRIPTION
**Pull request by AaronSpieler**
*Issue, if available:*
* sagemaker_sdk makefiles didnt work anymore
* test coverage is not very good

*Description of changes:*
* Fixing docker image makefiles  # due to sagemaker_container update
* Adding coverage of `.train(...)` method of GluonTSFramework
* Adding coverage for the fetching and cleaning up of the sage-maker dependencies

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AaronSpieler** [comment]
However, I'm not to sure about the Makefile tests. Essentially it could be that the `sagemaker_mxnet_container` library changes something and the test fails, even though no code has changed. It should rather raise a warning, but warning just get lost.

**Comment by AaronSpieler** [comment]
> However, I'm not to sure about the Makefile tests. Essentially it could be that the `sagemaker_mxnet_container` library changes something and the test fails, even though no code has changed. It should rather raise a warning, but warning just get lost.

Or it could be its own workflow? @jaheba

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=h1) Report
> Merging [#656](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/gluon-ts/commit/8013024d826fe4117057d8fe82d98c23ce423f37&el=desc) will **increase** coverage by `0.61%`.
> The diff coverage is `n/a`.


[Impacted file tree graph](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #656      +/-   ##
==========================================
+ Coverage   83.94%   84.56%   +0.61%     
==========================================
  Files         178      178              
  Lines       10337    10337              
==========================================
+ Hits         8677     8741      +64     
+ Misses       1660     1596      -64     
```

| [Impacted Files](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/gluonts/model/wavenet/\_network.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9kZWwvd2F2ZW5ldC9fbmV0d29yay5weQ==) | `98.81% <0.00%> (ø)` | :arrow_up: |
| [...uonts/monkey\_patch/monkey\_patch\_take\_along\_axis.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9ua2V5X3BhdGNoL21vbmtleV9wYXRjaF90YWtlX2Fsb25nX2F4aXMucHk=) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/trainer/learning\_rate\_scheduler.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvdHJhaW5lci9sZWFybmluZ19yYXRlX3NjaGVkdWxlci5weQ==) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/model/deepvar/\_network.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9kZWwvZGVlcHZhci9fbmV0d29yay5weQ==) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/transform/split.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvdHJhbnNmb3JtL3NwbGl0LnB5) | `98.24% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/model/deepar/\_network.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9kZWwvZGVlcGFyL19uZXR3b3JrLnB5) | `98.41% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/shell/sagemaker/path.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvc2hlbGwvc2FnZW1ha2VyL3BhdGgucHk=) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/model/seq2seq/\_transform.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9kZWwvc2VxMnNlcS9fdHJhbnNmb3JtLnB5) | `96.42% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/time\_feature/lag.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvdGltZV9mZWF0dXJlL2xhZy5weQ==) | `94.11% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/model/transformer/\_network.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9kZWwvdHJhbnNmb3JtZXIvX25ldHdvcmsucHk=) | `97.41% <0.00%> (ø)` | :arrow_up: |
| ... and [106 more](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree-more) | |

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=h1) Report
> Merging [#656](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/gluon-ts/commit/8013024d826fe4117057d8fe82d98c23ce423f37&el=desc) will **increase** coverage by `0.61%`.
> The diff coverage is `n/a`.


[Impacted file tree graph](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #656      +/-   ##
==========================================
+ Coverage   83.94%   84.56%   +0.61%     
==========================================
  Files         178      178              
  Lines       10337    10337              
==========================================
+ Hits         8677     8741      +64     
+ Misses       1660     1596      -64     
```

| [Impacted Files](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/gluonts/model/n\_beats/\_network.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbW9kZWwvbl9iZWF0cy9fbmV0d29yay5weQ==) | `90.90% <0.00%> (ø)` | :arrow_up: |
| [src/gluonts/dataset/common.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9jb21tb24ucHk=) | `82.65% <0.00%> (+2.04%)` | :arrow_up: |
| [src/gluonts/dataset/jsonl.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9qc29ubC5weQ==) | `84.84% <0.00%> (+3.03%)` | :arrow_up: |
| [src/gluonts/dataset/repository/datasets.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L2RhdGFzZXRzLnB5) | `86.48% <0.00%> (+10.81%)` | :arrow_up: |
| [...maker\_sdk/entry\_point\_scripts/train\_entry\_point.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbnVyc2VyeS9zYWdlbWFrZXJfc2RrL2VudHJ5X3BvaW50X3NjcmlwdHMvdHJhaW5fZW50cnlfcG9pbnQucHk=) | `75.51% <0.00%> (+48.97%)` | :arrow_up: |
| [src/gluonts/dataset/repository/\_m4.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L19tNC5weQ==) | `91.30% <0.00%> (+60.86%)` | :arrow_up: |
| [src/gluonts/dataset/repository/\_util.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L191dGlsLnB5) | `96.15% <0.00%> (+65.38%)` | :arrow_up: |

**Comment by AaronSpieler** [comment]
> However, I'm not to sure about the Makefile tests. Essentially it could be that the `sagemaker_mxnet_container` library changes something and the test fails, even though no code has changed. It should rather raise a warning, but warning just get lost.

This is fixed by the last commit. I marked the test as `xfail`. This means the test is expected to fail, so overall its failure will not make the whole test fail. Additionally, it will show up in the test report as `xfailed` or `xpassed` so we always know whether the makefile is broken.

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=h1) Report
> Merging [#656](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/gluon-ts/commit/47751d08f3f0211c78166e26652a11f116108837&el=desc) will **increase** coverage by `0.61%`.
> The diff coverage is `n/a`.


[Impacted file tree graph](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #656      +/-   ##
==========================================
+ Coverage   83.94%   84.56%   +0.61%     
==========================================
  Files         178      178              
  Lines       10337    10337              
==========================================
+ Hits         8677     8741      +64     
+ Misses       1660     1596      -64     
```

| [Impacted Files](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/gluonts/dataset/common.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9jb21tb24ucHk=) | `82.65% <0.00%> (+2.04%)` | :arrow_up: |
| [src/gluonts/dataset/jsonl.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9qc29ubC5weQ==) | `84.84% <0.00%> (+3.03%)` | :arrow_up: |
| [src/gluonts/dataset/repository/datasets.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L2RhdGFzZXRzLnB5) | `86.48% <0.00%> (+10.81%)` | :arrow_up: |
| [...maker\_sdk/entry\_point\_scripts/train\_entry\_point.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbnVyc2VyeS9zYWdlbWFrZXJfc2RrL2VudHJ5X3BvaW50X3NjcmlwdHMvdHJhaW5fZW50cnlfcG9pbnQucHk=) | `75.51% <0.00%> (+48.97%)` | :arrow_up: |
| [src/gluonts/dataset/repository/\_m4.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L19tNC5weQ==) | `91.30% <0.00%> (+60.86%)` | :arrow_up: |
| [src/gluonts/dataset/repository/\_util.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L191dGlsLnB5) | `96.15% <0.00%> (+65.38%)` | :arrow_up: |

File: test/nursery/sagemaker_sdk/test_makefiles.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        ),
+    ],
+)
+def test_make_pre_and_post_build_tasks(
+    docker_build_makefile_path, fetch_dependency_command
+):
+    # we need to write some data for this test, so we use a temporary directory
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_dir_path = Path(temp_dir)
+        copyfile(docker_build_makefile_path, str(temp_dir_path / "Makefile"))
```
Comment (jaheba) [inline test/nursery/sagemaker_sdk/test_makefiles.py]: Are you sure you need to apply `str` here?

File: test/nursery/sagemaker_sdk/test_makefiles.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            assert False, (
+                f"Something went wrong when fetching the sagemaker-mxnet-container dependency to building the "
+                f"container. Check sagemaker_sdk/cpu_training and run 'make' to debug. The original error message: {e.output} "
+            )
+
+        try:
+            subprocess.check_output(["make", "clean"], cwd=temp_dir_path)
+        except subprocess.CalledProcessError as e:
+            # The error message is not going to be useful in case of an error, so adding custom one:
+            assert False, (
```
Comment (jaheba) [inline test/nursery/sagemaker_sdk/test_makefiles.py]: I think raising an Error is cleaner here.

File: test/nursery/sagemaker_sdk/test_makefiles.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+from shutil import copyfile
+
+# Third-party imports
+import pytest
+
+# First-party imports
+import gluonts
+
+docker_images_path = (
+    Path(os.path.dirname(gluonts.__file__))
```
Comment (jaheba) [inline test/nursery/sagemaker_sdk/test_makefiles.py]: we don't need `os.path` here, the `Path` can do it itself

File: test/nursery/sagemaker_sdk/test_makefiles.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        ),
+    ],
+)
+def test_make_pre_and_post_build_tasks(
+    docker_build_makefile_path, fetch_dependency_command
+):
+    # we need to write some data for this test, so we use a temporary directory
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_dir_path = Path(temp_dir)
+        copyfile(docker_build_makefile_path, str(temp_dir_path / "Makefile"))
```
Comment (jaheba) [inline test/nursery/sagemaker_sdk/test_makefiles.py]:
```suggestion
        copyfile(docker_build_makefile_path, temp_dir_path / "Makefile")
```

File: test/nursery/sagemaker_sdk/test_makefiles.py (hunk 1 above)
Comment (AaronSpieler) [reply test/nursery/sagemaker_sdk/test_makefiles.py]: No we don't, you're right!

**Comment by alexw91** [comment]
# [Codecov](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=h1) Report
> Merging [#656](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=desc) into [master](https://codecov.io/gh/awslabs/gluon-ts/commit/a9c9f7ed2ba01765c0a56e238ae7aa0dbaf3314d&el=desc) will **increase** coverage by `0.61%`.
> The diff coverage is `n/a`.


[Impacted file tree graph](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #656      +/-   ##
==========================================
+ Coverage   83.98%   84.60%   +0.61%     
==========================================
  Files         178      178              
  Lines       10390    10390              
==========================================
+ Hits         8726     8790      +64     
+ Misses       1664     1600      -64     
```

| [Impacted Files](https://codecov.io/gh/awslabs/gluon-ts/pull/656?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/gluonts/dataset/common.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9jb21tb24ucHk=) | `82.65% <0.00%> (+2.04%)` | :arrow_up: |
| [src/gluonts/dataset/jsonl.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9qc29ubC5weQ==) | `84.84% <0.00%> (+3.03%)` | :arrow_up: |
| [src/gluonts/dataset/repository/datasets.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L2RhdGFzZXRzLnB5) | `86.48% <0.00%> (+10.81%)` | :arrow_up: |
| [...maker\_sdk/entry\_point\_scripts/train\_entry\_point.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvbnVyc2VyeS9zYWdlbWFrZXJfc2RrL2VudHJ5X3BvaW50X3NjcmlwdHMvdHJhaW5fZW50cnlfcG9pbnQucHk=) | `75.51% <0.00%> (+48.97%)` | :arrow_up: |
| [src/gluonts/dataset/repository/\_m4.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L19tNC5weQ==) | `91.30% <0.00%> (+60.86%)` | :arrow_up: |
| [src/gluonts/dataset/repository/\_util.py](https://codecov.io/gh/awslabs/gluon-ts/pull/656/diff?src=pr&el=tree#diff-c3JjL2dsdW9udHMvZGF0YXNldC9yZXBvc2l0b3J5L191dGlsLnB5) | `96.15% <0.00%> (+65.38%)` | :arrow_up: |

**Pull request merged**